Reduce allocations for one-sided usages of List.partition

retronym opened the pull request.

retronym committed:
Reduce allocations for one-sided usages of List.partition

retronym committed:
Reuse an instance of a TypeCollector used in hot loop in LUBs

diesalbla commented on src/reflect/scala/reflect/internal/util/Collections.scala:
```diff
@@ -363,6 +363,50 @@ trait Collections {
+        if (p0) (xs, Nil) else (Nil, xs)
+      } else {
+        val r = (ayes.toList, nays.toList)
+        println(r)
```
I think this should be removed?

diesalbla commented on src/reflect/scala/reflect/internal/util/Collections.scala:
```diff
@@ -363,6 +363,50 @@ trait Collections {
   }
 
+  private val TupleOfNil = (Nil, Nil)
+  final def partitionConserve[A](xs: List[A])(pred: A => Boolean): (List[A], List[A]) = {
```
This looks like the same code added to the `List` class... could it be a firts draft, or is there a difference between them?

retronym replied:
Oops!

som-snytt replied:
`s/firts draft/farts draft/`

sjrd commented on src/library/scala/collection/immutable/List.scala:
```diff
@@ -556,6 +556,44 @@ sealed abstract class List[+A]
+  override def partition(p: A => Boolean): (List[A], List[A]) = {
+    if (isEmpty) List.TupleOfNil
+    else {
+      val b0 = p(this.head)
```
Having calls to `p` in 2 different places is problematic for closure elimination in Scala.js. Could you instead initialize `b0` with a constant, and set it to `b` after initializing `b` if `n == 0`?

lrytz commented on src/library/scala/collection/immutable/List.scala:
```diff
@@ -556,6 +556,44 @@ sealed abstract class List[+A]
+            prefix += zs.head
+            zs = zs.tail
+            j += 1
+          }
```
In the extreme case (if only the last element of the list needs to go to the other bucket), the entire list is traversed twice. Is that a good trade off?

diesalbla replied:
@lrytz Currently, the class `List` inherits its implementation of `partition` from the [`Iterable` trait](https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/Iterable.scala#L418-L430) (or at least I could not find any `override` in between). The Scaladoc comment for `Iterable.partition` says: 

>   The default implementation provided here needs to traverse the collection twice.
>   Strict collections have an overridden version of `partition` in `StrictOptimizedIterableOps`,
>  which requires only a single traversal.

So this is not a loss.

retronym replied:
List does inherit `StringOptimizedIterableOps.partition`:

```
Welcome to Scala 2.13.1 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_172).
Type in expressions for evaluation. Or try :help.

scala> :power
Power mode enabled. :phase is at typer.
import scala.tools.nsc._, intp.global._, definitions._
Try :help or completions for vals._ and power._

scala> typeOf[List[_]].member(TermName("partition")).fullLocationString
res0: String = method partition in trait StrictOptimizedIterableOps
```

retronym replied:
I haven't been able to measure the performance slowdown. `ListBenchmark.partition_exc_last` tests this case. I tried with `-psize=1000000`, as for smaller data sizes the second traversal is often cheaper due to CPU cache effects.

lrytz replied:
Thanks. Still slightly surprised - do you have an intuition?

diesalbla replied:
The difference for that worst case is between one traversal and allocating `N` new `::` objects, or two traversals and allocating `N-1` objects. It may be that the cost of the extra traversal is negligible next to that of the allocations.

To put it in another way, for each element we have: 
- Before: load  `::` object, load element, apply function to element, create new `::` object.
- After: load `::` object, load element, apply function to element; and for the second iteration (clone prefix), re-load `::` object, and create new `::` object. 

It is only the second re-loading of the `::` object, per iteration, (from the source list) that changes from one method to the other. My intuition would be, this is negligible next to the rest.

sjrd commented on src/library/scala/collection/immutable/List.scala:
```diff
@@ -556,6 +556,45 @@ sealed abstract class List[+A]
+        if (!canConserve)
+          (if (b) ayes else nays) += y
+        n += 1
+        ys = ys.tail
```
May I suggest the following rearrangement, where most of the loop body is gated by an initial `if (canConserve)`:
```scala
      while (!ys.isEmpty) {
        val y = ys.head
        val b = p(y)
        if (canConserve) {
          if (n == 0) b0 = b
          else if (b != b0) {
            canConserve = false
            ayes = new ListBuffer[A]
            nays = new ListBuffer[A]
            val prefix = if (b0) ayes else nays
            var j = 0
            var zs = this
            while (j < n) {
              prefix += zs.head
              zs = zs.tail
              j += 1
            }
            (if (b) ayes else nays) += y
          }
          n += 1
        } else {
          (if (b) ayes else nays) += y
        }
        ys = ys.tail
      }
```

som-snytt commented:
Sometimes I wonder what retronym is doing right now, like is he watching wildfire coverage?

retronym commented:
I've dismissed the positive reviews as I found my mistake with running the benchmarks: running without the optimizer enabled ended up disadvantaging the status quo which calls `Growable.+=` inside a loop which doesn't always JIT inline (a repeat of https://github.com/scala/bug/issues/11627).

But, running with the Scala optimizer enabled requires caution to do clean builds when editing implementations as SBT's incremental builds can undercompile with the optimizer enabled.

```
[info] ListBenchmark.partition_exc_last_inherited                                   1000000  avgt    8   8608980.284 ±  644398.411   ns/op
[info] ListBenchmark.partition_exc_last_new                               1000000  avgt    8  11752659.893 ± 1251025.740   ns/op
```

So we get the expected slowdown in this worst-case data input to the benchmark.

I also tried copying the inherited method implementation into `List` and removing indirection through `Iterator` in favour of a while loop. This gave:

```
[info] ListBenchmark.partition_exc_last_while                                   1000000  avgt    8   7072744.769 ±  678075.766   ns/op
```

There are a few alternatives we could consider here:

  - Accept the slowdown as a valid tradeoff against the reduced allocation rate and possible reduced retained memory size in applications due to structural sharing
  - Return to a one-pass algorithm that implements structural sharing as post-processing step.
     - But, keep `partitionConserve` in the compiler `Collections` utilities and use it in the hot code in the LUB calculations.

diesalbla commented:
> I also tried copying the inherited method implementation into List and removing indirection through Iterator in favour of a while loop. 

If this improves in all scenarios, it could be good enough.

> - [...] keep partitionConserve in the compiler Collections utilities and use it in the hot code in the LUB calculations.

We could do that, even if the optimisation cannot be useful for all cases.

retronym committed:
Demote new List.partition implenentation to a compiler util

  - Use it is a few performance sensitive call sites in the compiler
  - Retain the override of List.partition, but change it to a one-pass
    algorithm does post-factum structural sharing.

som-snytt commented:
Some of us would consider that a promotion.
